### DISCUSSION TITLE
Using React Flow event handlers directly from custom component?

### DISCUSSION THREAD

**Discussion by kahboom** [xyflow/xyflow#1954] [Q&A] [answered]
Hello, I'm using more than one custom component and I need for nodes in one of them to be both clickable _and_ draggable (click and drag execute different functions). I have no problem with `onClick`, but no other native event handlers seem to be supported. My questions are:

1. I need to disable `onClick`, but only when the user is dragging, similar to [this issue](https://github.com/wbkd/react-flow/discussions/1412). This is easy to do when you only have one type of custom node, because you can just specify it at the main `<ReactFlow/>` component using the [component event handlers](https://reactflow.dev/docs/api/component-props/). I have tried to use it for `onDrag` and other native event handlers, but it's not being executed. Does anyone know if this is possible?
2. Is it possible to use the event handlers provided by React Flow in the docs [here](https://reactflow.dev/docs/api/component-props/), but in the custom component, so that we can use drag and click separately like in the example [here](https://reactflow.dev/examples/custom-node/)? Or even just browser native event handlers like `onDragStart`.

**Comment by moklick** [comment, marked as answer]
Hey @kahboom. The `onElementClick` handler should only be fired when you click a node (can be a custom node as well) and not when you drag it.

**Reply by kahboom** [reply]
Hey @moklick , thanks for the quick reply. It's strange, I wasn't able to reproduce it on this [CodeSandbox](https://codesandbox.io/s/react-flow-custom-node-dxc65s?file=/src/CustomNodeFlow.tsx), but a click is definitely being triggered. [This](https://github.com/KaotoIO/kaoto-ui/blob/main/src/components/Visualization.tsx#L132) is the main component where I start building the elements, and [this](https://github.com/KaotoIO/kaoto-ui/blob/main/src/components/VisualizationStep.tsx#L64) is the custom component. I will keep checking and if I find the bug I'll post on here in case it helps anyone else.

**Reply by moklick** [reply]
You don't need to add an `onClick` handler within your custom node. Every node gets wrapped by React Flow and the `onElementClick` handler that you pass to the `<ReactFlow />` component is used.

**Reply by kahboom** [reply]
I had to add `onClick` handler to my custom node because there is more than one thing to click on in the custom node and I didn't know how to get that context. For example, if you click the + button [here](https://github.com/KaotoIO/kaoto-ui/blob/main/src/components/VisualizationStep.tsx#L58) it triggers a popover (to add another element/node), but clicking on the element/node should open a separate dialog that shows the data for the node. Do you have any idea for how to approach this in a better way?

**Reply by moklick** [reply]
For the elements inside the node that have independent event handlers, you need to add the class name `nodrag`: https://codesandbox.io/s/react-flow-custom-node-forked-7hzvzr?file=/src/ColorSelectorNode.tsx:738-756 . Check the "log" button.

**Reply by kahboom** [reply]
Thanks @moklick this seems to be the best solution.
